RetainPtr’s leakRef() and move constructors are incorrect when ARC is enabled

geoffreygaren opened the pull request:
#### 0fd9dfa52fae0568f15f0922ff1c470369475cf7
<pre>
RetainPtr’s leakRef() and move constructors are incorrect when ARC is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=288301">https://bugs.webkit.org/show_bug.cgi?id=288301</a>
&lt;<a href="https://rdar.apple.com/problem/145393261">rdar://problem/145393261</a>&gt;

Reviewed by Timothy Hatcher and David Kilzer.

They net out +2 or -1, so you either get a leak or a crash (unless you combine
them with adoptRef() or bridge_cast(), which also incorrectly net out -1, in
which case it all balances out, lol).

When ARC is enabled, converting between id and CFTypeRef is fraught. The
compiler accepts any conversion you type, but most conversions implicitly
forget or hallucinate a refcount.

To solve this, tease apart id and CFTypeRef so that ARC code never needs to
convert. id gets stored as id and CFTypeRef gets stored as CFTypeRef.

One implicit conversion remains: C++ files type pun id to CFTypeRef, so you can
initialize an id in an ObjC file and move, copy, or destroy it in a C++ file.
To prevent ARC confusion, there’s no cast; instead we #ifdef rename the declared
type of the field when compiling for C++. This is valid because id and CFTypeRef
are ABI equivalent.

This approach ends up tightening type-safety enforcement, since we no longer
cast m_ptr when we load/store it. As a result, C++ code that uses id needs to
stop doing that, since id is not a usable type in C++. Also, tests need to stop
using RetainPtr&lt;NSString*&gt;, and standardize on RetainPtr&lt;NSString&gt;.

Also fixed bridge_cast to be +0 in ARC, since it can no longer rely on leakRef()
being +2.

Also fixed TypeCastsCocoaARC to use more autorelease pools in order to get
an accurate refcount.

Now that this all works, we can enable the previously failing leakRef() API test. Yay!

* Source/WTF/wtf/RetainPtr.h:
(retainFoundationPtr):
(releaseFoundationPtr):
(autoreleaseFoundationPtr):
(WTF::RetainPtr::RetainPtr):
(WTF::RetainPtr::isHashTableDeletedValue const):
(WTF::RetainPtr::get const):
(WTF::RetainPtr::operator-&gt; const):
(WTF::RetainPtr::operator PtrType const):
(WTF::RetainPtr::hashTableDeletedValue):
(WTF::RetainPtr&lt;T&gt;::~RetainPtr):
(WTF::RetainPtr&lt;T&gt;::RetainPtr):
(WTF::RetainPtr&lt;T&gt;::clear):
(WTF::RetainPtr&lt;T&gt;::autorelease):
(WTF::adoptCF):
(WTF::adoptNS):
(WTF::RetainPtr::checkType): Deleted.
(WTF::RetainPtr::fromStorageTypeHelper): Deleted.
(WTF::RetainPtr::fromStorageType): Deleted.
(WTF::RetainPtr::toStorageType): Deleted.
* Source/WebCore/bridge/objc/objc_runtime.h:
* Source/WebCore/platform/PlatformSpeechSynthesisUtterance.h:
* Source/WebCore/platform/network/cf/AuthenticationChallenge.h:
* Tools/DumpRenderTree/AccessibilityUIElement.cpp:
(getIsValidCallback):
(AccessibilityUIElement::makeJSAccessibilityUIElement):
(AccessibilityUIElement::isEqual): Deleted.
* Tools/DumpRenderTree/AccessibilityUIElement.h:
(AccessibilityUIElement::hasPlatformUIElement const):
* Tools/DumpRenderTree/ios/AccessibilityUIElementIOS.mm:
(AccessibilityUIElement::isEqual):
* Tools/DumpRenderTree/mac/AccessibilityUIElementMac.mm:
(AccessibilityUIElement::isEqual):
* Tools/TestWebKitAPI/Tests/WTF/ns/RetainPtr.mm:
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, ConstructionFromMutableNSType)):
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, ConstructionFromSimilarNSType)):
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, ConstructionFromSimilarNSTypeReversed)):
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, MoveAssignmentFromSimilarNSType)):
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, MoveAssignmentFromSimilarNSTypeReversed)):
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, LeakRef)):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarker.cpp:
(WTR::AccessibilityTextMarker::platformTextMarker const): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarker.h:
(WTR::AccessibilityTextMarker::platformTextMarker const):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.cpp:
(WTR::AccessibilityTextMarkerRange::platformTextMarkerRange const): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.h:
(WTR::AccessibilityTextMarkerRange::platformTextMarkerRange const):

Canonical link: <a href="https://commits.webkit.org/290928@main">https://commits.webkit.org/290928@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f740c068b5ce6f47e8f540e27e7739238d4887c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91465 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10997 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96434 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42153 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93515 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11374 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19426 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70229 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27750 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94466 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8673 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82860 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50555 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8438 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/447 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41323 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84269 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78758 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/452 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98437 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90214 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18627 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79251 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18882 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78698 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78455 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22980 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/346 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11762 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14477 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18625 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23901 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112798 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18335 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32709 "Found 13 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.default, microbenchmarks/memcpy-wasm-medium.js.dfg-eager, microbenchmarks/memcpy-wasm-medium.js.no-llint, microbenchmarks/memcpy-wasm-small.js.dfg-eager, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-slow-memory, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-collect-continuously, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-eager, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.default-wasm, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.wasm-collect-continuously ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21795 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20101 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WTF/wtf/RetainPtr.h:
```diff
@@ -140,44 +167,22 @@ template<typename T> class RetainPtr {
-    static constexpr CFTypeRef toStorageType(id ptr) { return (__bridge CFTypeRef)ptr; }
-    static constexpr CFTypeRef toStorageType(CFTypeRef ptr) { return ptr; }
+#if __has_feature(objc_arc)
+    // ARC will try to retain/release this value, but it looks like a tagged immediate, so retain/release ends up being a no-op -- see _objc_isTaggedPointer() <objc-internal.h>.
```
Nit:  Change to "... in < objc-internal.h>."

```suggestion
    // ARC will try to retain/release this value, but it looks like a tagged immediate, so retain/release ends up being a no-op -- see _objc_isTaggedPointer() in <objc-internal.h>.
```

ddkilzer commented on Source/WTF/wtf/RetainPtr.h:
```diff
@@ -65,9 +65,30 @@ typedef struct objc_object *id;
+inline void releaseFoundationPtr(CFTypeRef ptr) { CFRelease(ptr); }
+#if HAVE(CFAUTORELEASE)
+inline void autoreleaseFoundationPtr(CFTypeRef ptr) { CFAutorelease(ptr); }
+#endif
```
Nit: Separate from this patch, I think we can probably remove `HAVE(CFAUTORELEASE)` because it only existed for the Apple Windows port.

ddkilzer commented on Source/WTF/wtf/RetainPtr.h:
```diff
@@ -65,9 +65,30 @@ typedef struct objc_object *id;
+inline void releaseFoundationPtr(id ptr) { [ptr release]; }
+inline void autoreleaseFoundationPtr(id ptr) { [ptr autorelease]; }
+#endif
+#endif
```
Would be nice if these were private `static` methods on `RetainPtr<>` since it would be unexpected to use them outside of the class.

ddkilzer reviewed (approved):
r=me, but I'd like to see the visibility of the helper functions scoped to private functions of `RetainPtr<>`.  (Nits not required to fix before landing.)

geoffreygaren committed:
RetainPtr’s leakRef() and move constructors are incorrect when ARC is enabled
https://bugs.webkit.org/show_bug.cgi?id=288301
<rdar://problem/145393261>

Reviewed by Timothy Hatcher and David Kilzer.

They net out +2 or -1, so you either get a leak or a crash (unless you combine
them with adoptRef() or bridge_cast(), which also incorrectly net out -1, in
which case it all balances out, lol).

When ARC is enabled, converting between id and CFTypeRef is fraught. The
compiler accepts any conversion you type, but most conversions implicitly
forget or hallucinate a refcount.

To solve this, tease apart id and CFTypeRef so that ARC code never needs to
convert. id gets stored as id and CFTypeRef gets stored as CFTypeRef.

One implicit conversion remains: C++ files type pun id to CFTypeRef, so you can
initialize an id in an ObjC file and move, copy, or destroy it in a C++ file.
To prevent ARC confusion, there’s no cast; instead we #ifdef rename the declared
type of the field when compiling for C++. This is valid because id and CFTypeRef
are ABI equivalent.

This approach ends up tightening type-safety enforcement, since we no longer
cast m_ptr when we load/store it. As a result, C++ code that uses id needs to
stop doing that, since id is not a usable type in C++. Also, tests need to stop
using RetainPtr<NSString*>, and standardize on RetainPtr<NSString>.

Also fixed bridge_cast to be +0 in ARC, since it can no longer rely on leakRef()
being +2.

Also fixed TypeCastsCocoaARC to use more autorelease pools in order to get
an accurate refcount.

Now that this all works, we can enable the previously failing leakRef() API test. Yay!

* Source/WTF/wtf/RetainPtr.h:
(retainFoundationPtr):
(releaseFoundationPtr):
(autoreleaseFoundationPtr):
(WTF::RetainPtr::RetainPtr):
(WTF::RetainPtr::isHashTableDeletedValue const):
(WTF::RetainPtr::get const):
(WTF::RetainPtr::operator-> const):
(WTF::RetainPtr::operator PtrType const):
(WTF::RetainPtr::hashTableDeletedValue):
(WTF::RetainPtr<T>::~RetainPtr):
(WTF::RetainPtr<T>::RetainPtr):
(WTF::RetainPtr<T>::clear):
(WTF::RetainPtr<T>::autorelease):
(WTF::adoptCF):
(WTF::adoptNS):
(WTF::RetainPtr::checkType): Deleted.
(WTF::RetainPtr::fromStorageTypeHelper): Deleted.
(WTF::RetainPtr::fromStorageType): Deleted.
(WTF::RetainPtr::toStorageType): Deleted.
* Source/WebCore/bridge/objc/objc_runtime.h:
* Source/WebCore/platform/PlatformSpeechSynthesisUtterance.h:
* Source/WebCore/platform/network/cf/AuthenticationChallenge.h:
* Tools/DumpRenderTree/AccessibilityUIElement.cpp:
(getIsValidCallback):
(AccessibilityUIElement::makeJSAccessibilityUIElement):
(AccessibilityUIElement::isEqual): Deleted.
* Tools/DumpRenderTree/AccessibilityUIElement.h:
(AccessibilityUIElement::hasPlatformUIElement const):
* Tools/DumpRenderTree/ios/AccessibilityUIElementIOS.mm:
(AccessibilityUIElement::isEqual):
* Tools/DumpRenderTree/mac/AccessibilityUIElementMac.mm:
(AccessibilityUIElement::isEqual):
* Tools/TestWebKitAPI/Tests/WTF/ns/RetainPtr.mm:
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, ConstructionFromMutableNSType)):
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, ConstructionFromSimilarNSType)):
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, ConstructionFromSimilarNSTypeReversed)):
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, MoveAssignmentFromSimilarNSType)):
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, MoveAssignmentFromSimilarNSTypeReversed)):
(TestWebKitAPI::TEST(RETAIN_PTR_TEST_NAME, LeakRef)):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarker.cpp:
(WTR::AccessibilityTextMarker::platformTextMarker const): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarker.h:
(WTR::AccessibilityTextMarker::platformTextMarker const):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.cpp:
(WTR::AccessibilityTextMarkerRange::platformTextMarkerRange const): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.h:
(WTR::AccessibilityTextMarkerRange::platformTextMarkerRange const):

Canonical link: https://commits.webkit.org/290928@main
